### ISSUE TITLE
📝 Index parameters are detected as Type parameters

### ISSUE THREAD

**Issue by johnpalit-hotjar** [biomejs/biome#7532]
### Environment information

```bash
CLI:
  Version:                      2.2.2
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v22.18.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         yarn/4.9.3

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true

Formatter:
  Format with errors:           false
  Indent style:                 Space
  Indent width:                 2
  Line ending:                  LF
  Line width:                   120
  Attribute position:           Auto
  Bracket spacing:              true
  Includes:                     **

JavaScript Formatter:
  Enabled:                      unset
  JSX quote style:              Double Quotes
  Quote properties:             As needed
  Trailing commas:              All
  Semicolons:                   Always
  Arrow parentheses:            Always
  Bracket spacing:              true
  Bracket same line:            false
  Quote style:                  Single Quotes
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Attribute position:           Auto

JSON Formatter:
  Enabled:                      unset
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Trailing Commas:              unset
  Expand lists:                 unset

CSS Formatter:
  Enabled:                      unset
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Quote style:                  unset

GraphQL Formatter:
  Enabled:                      unset
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Bracket spacing:              unset
  Quote style:                  unset

Workspace:
  Open Documents:               0
```

### Configuration

```JSON
{
  "$schema": "https://biomejs.dev/schemas/2.2.2/schema.json",
  "linter": {
    "enabled": true,
    "rules": {
      "style": {
        "useNamingConvention": {
          "level": "error",
          "options": {
            "conventions": [
              {
                "selector": {
                  "kind": "typeParameter"
                },
                "formats": ["PascalCase"]
              },
              {
                "selector": {
                  "kind": "indexParameter"
                },
                "formats": ["camelCase"]
              }
            ]
          }
        }
      }
    }
  }
}
```

### Playground link

https://biomejs.dev/playground/?indentStyle=space&quoteStyle=single&files.biome.json=ewAKACAAIAAiACQAcwBjAGgAZQBtAGEAIgA6ACAAIgBoAHQAdABwAHMAOgAvAC8AYgBpAG8AbQBlAGoAcwAuAGQAZQB2AC8AcwBjAGgAZQBtAGEAcwAvADIALgAyAC4AMgAvAHMAYwBoAGUAbQBhAC4AagBzAG8AbgAiACwACgAgACAAIgBsAGkAbgB0AGUAcgAiADoAIAB7AAoAIAAgACAAIAAiAGUAbgBhAGIAbABlAGQAIgA6ACAAdAByAHUAZQAsAAoAIAAgACAAIAAiAHIAdQBsAGUAcwAiADoAIAB7AAoAIAAgACAAIAAgACAAIgBzAHQAeQBsAGUAIgA6ACAAewAKACAAIAAgACAAIAAgACAAIAAiAHUAcwBlAE4AYQBtAGkAbgBnAEMAbwBuAHYAZQBuAHQAaQBvAG4AIgA6ACAAewAKACAAIAAgACAAIAAgACAAIAAgACAAIgBsAGUAdgBlAGwAIgA6ACAAIgBlAHIAcgBvAHIAIgAsAAoAIAAgACAAIAAgACAAIAAgACAAIAAiAG8AcAB0AGkAbwBuAHMAIgA6ACAAewAKACAAIAAgACAAIAAgACAAIAAgACAAIAAgACIAYwBvAG4AdgBlAG4AdABpAG8AbgBzACIAOgAgAFsACgAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAewAKACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIgBzAGUAbABlAGMAdABvAHIAIgA6ACAAewAKACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACIAawBpAG4AZAAiADoAIAAiAHQAeQBwAGUAUABhAHIAYQBtAGUAdABlAHIAIgAKACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAfQAsAAoAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAiAGYAbwByAG0AYQB0AHMAIgA6ACAAWwAiAFAAYQBzAGMAYQBsAEMAYQBzAGUAIgBdAAoAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgAH0ALAAKACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAB7AAoAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAiAHMAZQBsAGUAYwB0AG8AcgAiADoAIAB7AAoAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIgBrAGkAbgBkACIAOgAgACIAaQBuAGQAZQB4AFAAYQByAGEAbQBlAHQAZQByACIACgAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgAH0ALAAKACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAIgBmAG8AcgBtAGEAdABzACIAOgAgAFsAIgBjAGEAbQBlAGwAQwBhAHMAZQAiAF0ACgAgACAAIAAgACAAIAAgACAAIAAgACAAIAAgACAAfQAKACAAIAAgACAAIAAgACAAIAAgACAAIAAgAF0ACgAgACAAIAAgACAAIAAgACAAIAAgAH0ACgAgACAAIAAgACAAIAAgACAAfQAKACAAIAAgACAAIAAgAH0ACgAgACAAIAAgAH0ACgAgACAAfQAKAH0A&files.main.ts=ZQBuAHUAbQAgAFMAbwBtAGUATABpAHMAdAAgAHsACgAgACAASQB0AGUAbQBPAG4AZQAgAD0AIAAnAGkAdABlAG0ATwBuAGUAJwAsAAoAIAAgAEkAdABlAG0AVAB3AG8AIAA9ACAAJwBpAHQAZQBtAFQAdwBvACcALAAKACAAIABJAHQAZQBtAFQAaAByAGUAZQAgAD0AIAAnAGkAdABlAG0AVABoAHIAZQBlACcALAAKAH0ACgAKAHQAeQBwAGUAIABJAG4AZABlAHgAZQBkAEIAeQBFAG4AdQBtACAAPQAgAHsAIABbAGsAZQB5ACAAaQBuACAAUwBvAG0AZQBMAGkAcwB0AF0AOgAgAHMAdAByAGkAbgBnACAAfQA7AAoACgBlAHgAcABvAHIAdAAgAGMAbwBuAHMAdAAgAGUAeABhAG0AcABsAGUAOgAgAEkAbgBkAGUAeABlAGQAQgB5AEUAbgB1AG0AIAA9ACAAewAKACAAIABpAHQAZQBtAE8AbgBlADoAIAAnAGkAdABlAG0AIABvAG4AZQAnACwACgAgACAAaQB0AGUAbQBUAHcAbwA6ACAAJwBpAHQAZQBtACAAdAB3AG8AJwAsAAoAIAAgAGkAdABlAG0AVABoAHIAZQBlADoAIAAnAGkAdABlAG0AIAB0AGgAcgBlAGUAJwAsAAoAfQA7AA%3D%3D

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by johnpalit-hotjar** [comment]
If `typeParameter` naming convention is set to `PascalCase` (which I believe is the default) then index parameters are marked as errors if they do not match this naming convention. This is the case even if `indexParameter` naming convention config is also set, e.g. to `camelCase`, which should be the default.

This is confusing, as e.g. in:

```ts
enum SomeList {
  ItemOne = 'itemOne',
  ItemTwo = 'itemTwo',
  ItemThree = 'itemThree',
}

type IndexedByEnum = { [key in SomeList]: string };

export const example: IndexedByEnum = {
  itemOne: 'item one',
  itemTwo: 'item two',
  itemThree: 'item three',
};
```

I'd expect to be able to use `key` not `Key` in the index, but instead get the error:
```
/main.ts:7:25 [lint/style/useNamingConvention](https://biomejs.dev/linter/rules/use-naming-convention) ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ This type parameter name should be in PascalCase.
  
    5 │ }
    6 │ 
  > 7 │ type IndexedByEnum = { [key in SomeList]: string };
      │                         ^^^
    8 │ 
    9 │ export const example: IndexedByEnum = {
```

I've provided a cut down `biome.json` to highlight the relevant config (which I believe to be the same as the defaults, just wanted to show the relevant parts). I used this config on the playground to verify that the same behaviour occurs there, without some side effect on my local setup (which it does).

**Comment by Conaclos** [comment]
`{ [key in SomeList]: string }` is not an index parameter, it is a mapped type.
This is why `indexParameter` has no effet on that. `key` is recognized as a type parameter because it is a type.

Taking a look at the [TS docs about mapped types](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html), it seems that they use `PascalCase` naming convention. Could you provide some resources that support your case?

A possibility is to introduce a dedicated selector like `mappedTypeParameter` and to allow both `camelCase` and `PascalCase`.

**Comment by arendjr** [comment]
It looks like @Conaclos 's interpretation is indeed correct, and since there's not been any request for introducing a new option, I think this one is safe to close.
